File: src/Lexer.cpp
#include "Idot/Lexer.h"

#include <cctype>
#include <unordered_map>

#include "Idot/Diagnostics.h"

namespace idot {
namespace {

const std::unordered_map<std::string, TokenType> kKeywords = {
    {"let", TokenType::KeywordLet},   {"if", TokenType::KeywordIf},
    {"else", TokenType::KeywordElse}, {"true", TokenType::KeywordTrue},
    {"false", TokenType::KeywordFalse}, {"nil", TokenType::KeywordNil},
    {"print", TokenType::KeywordPrint},
};

bool IsAlpha(char value) { return std::isalpha(static_cast<unsigned char>(value)) != 0 || value == '_'; }

bool IsAlphaNumeric(char value) {
  return std::isalnum(static_cast<unsigned char>(value)) != 0 || value == '_';
}

}  // namespace

Lexer::Lexer(std::string source) : source_(std::move(source)) {}

std::vector<Token> Lexer::ScanTokens() {
  while (!IsAtEnd()) {
    start_ = current_;
    startLine_ = line_;
    startColumn_ = column_;
    ScanToken();
  }

  tokens_.push_back(Token{TokenType::EndOfFile, "", line_, column_});
  return tokens_;
}

bool Lexer::IsAtEnd() const { return current_ >= source_.size(); }

char Lexer::Advance() {
  char value = source_[current_++];
  if (value == '\n') {
    ++line_;
    column_ = 1;
  } else {
    ++column_;
  }
  return value;
}

bool Lexer::Match(char expected) {
  if (IsAtEnd()) {
    return false;
  }
  if (source_[current_] != expected) {
    return false;
  }
  ++current_;
  ++column_;
  return true;
}

char Lexer::Peek() const { return IsAtEnd() ? '\0' : source_[current_]; }

char Lexer::PeekNext() const {
  if (current_ + 1 >= source_.size()) {
    return '\0';
  }
  return source_[current_ + 1];
}

void Lexer::ScanToken() {
  const char token = Advance();
  switch (token) {
    case '(':
      AddToken(TokenType::LeftParen);
      break;
    case ')':
      AddToken(TokenType::RightParen);
      break;
    case '{':
      AddToken(TokenType::LeftBrace);
      break;
    case '}':
      AddToken(TokenType::RightBrace);
      break;
    case ',':
      AddToken(TokenType::Comma);
      break;
    case '.':
      AddToken(TokenType::Dot);
      break;
    case '-':
      AddToken(TokenType::Minus);
      break;
    case '+':
      AddToken(TokenType::Plus);
      break;
    case ';':
      AddToken(TokenType::Semicolon);
      break;
    case '*':
      AddToken(TokenType::Star);
      break;
    case '%':
      AddToken(TokenType::Percent);
      break;
    case '!':
      AddToken(Match('=') ? TokenType::BangEqual : TokenType::Bang);
      break;
    case '=':
      AddToken(Match('=') ? TokenType::EqualEqual : TokenType::Equal);
      break;
    case '<':
      AddToken(Match('=') ? TokenType::LessEqual : TokenType::Less);
      break;
    case '>':
      AddToken(Match('=') ? TokenType::GreaterEqual : TokenType::Greater);
      break;
    case '/':
      if (Match('/')) {
        while (Peek() != '\n' && !IsAtEnd()) {
          Advance();
        }
      } else {
        AddToken(TokenType::Slash);
      }
      break;
    case '"':
      String();
      break;
    case ' ':
    case '\r':
    case '\t':
    case '\n':
      break;
    default:
      if (std::isdigit(static_cast<unsigned char>(token)) != 0) {
        Number();
      } else if (IsAlpha(token)) {
        Identifier();
      } else {
        throw DiagnosticError(ErrorPhase::Lex, startLine_, startColumn_,
                              std::string("Unexpected character '") + token + "'.");
      }
      break;
  }
}

void Lexer::AddToken(TokenType type) { AddToken(type, source_.substr(start_, current_ - start_)); }

void Lexer::AddToken(TokenType type, const std::string& lexeme) {
  tokens_.push_back(Token{type, lexeme, startLine_, startColumn_});
}

void Lexer::String() {
  while (Peek() != '"' && !IsAtEnd()) {
    Advance();
  }

  if (IsAtEnd()) {
    throw DiagnosticError(ErrorPhase::Lex, startLine_, startColumn_, "Unterminated string.");
  }

  Advance();
  AddToken(TokenType::String, source_.substr(start_ + 1, current_ - start_ - 2));
}

void Lexer::Number() {
  while (std::isdigit(static_cast<unsigned char>(Peek())) != 0) {
    Advance();
  }

  if (Peek() == '.' && std::isdigit(static_cast<unsigned char>(PeekNext())) != 0) {
    Advance();
    while (std::isdigit(static_cast<unsigned char>(Peek())) != 0) {
      Advance();
    }
  }

  AddToken(TokenType::Number);
}

void Lexer::Identifier() {
  while (IsAlphaNumeric(Peek())) {
    Advance();
  }

  const std::string text = source_.substr(start_, current_ - start_);
  auto keyword = kKeywords.find(text);
  if (keyword != kKeywords.end()) {
    AddToken(keyword->second, text);
    return;
  }
  AddToken(TokenType::Identifier, text);
}

}  // namespace idot

File: src/lexer.rs
#[derive(Debug, Clone, PartialEq)]
pub enum Token {
    Integer(i64),
    Identifier(String),

    Fn,
    Let,
    If,
    Else,
    While,
    Return,
    True,
    False,
    Int,
    Bool,
    Print,

    Plus,
    Minus,
    Star,
    Slash,

    Equal,
    EqualEqual,
    BangEqual,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,

    LParen,
    RParen,
    LBrace,
    RBrace,
    Comma,
    Semicolon,
    Colon,
    Arrow,

    EOF,
    Illegal(String),
}

pub struct Lexer {
    chars: Vec<char>,
    pos: usize,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        Lexer {
            chars: input.chars().collect(),
            pos: 0,
        }
    }

    fn peek(&self) -> Option<char> {
        self.chars.get(self.pos).copied()
    }

    fn advance(&mut self) -> Option<char> {
        let ch = self.chars.get(self.pos).copied();
        self.pos += 1;
        ch
    }

    fn skip_whitespace(&mut self) {
        while let Some(ch) = self.peek() {
            if ch == '/' && self.pos + 1 < self.chars.len() && self.chars[self.pos + 1] == '/' {
                while let Some(c) = self.peek() {
                    if c == '\n' || c == '\r' {
                        break;
                    }
                    self.pos += 1;
                }
            }
            if !ch.is_ascii_whitespace() {
                break;
            }
            self.pos += 1;
        }
    }

    pub fn tokenize(&mut self) -> Vec<Token> {
        let mut tokens = Vec::new();

        loop {
            self.skip_whitespace();
            match self.peek() {
                None => {
                    tokens.push(Token::EOF);
                    break;
                }
                Some(ch) if ch.is_ascii_digit() => {
                    let mut num = String::new();
                    while let Some(d) = self.peek() {
                        if d.is_ascii_digit() {
                            num.push(d);
                            self.pos += 1;
                        } else {
                            break;
                        }
                    }
                    tokens.push(Token::Integer(num.parse().unwrap()));
                }
                Some(ch) if ch.is_ascii_alphabetic() || ch == '_' => {
                    let mut ident = String::new();
                    while let Some(a) = self.peek() {
                        if a.is_ascii_alphanumeric() || a == '_' {
                            ident.push(a);
                            self.pos += 1;
                        } else {
                            break;
                        }
                    }
                    tokens.push(match ident.as_str() {
                        "fn" => Token::Fn,
                        "let" => Token::Let,
                        "if" => Token::If,
                        "else" => Token::Else,
                        "while" => Token::While,
                        "return" => Token::Return,
                        "true" => Token::True,
                        "false" => Token::False,
                        "int" => Token::Int,
                        "bool" => Token::Bool,
                        "print" => Token::Print,
                        _ => Token::Identifier(ident),
                    });
                }
                Some('=') => {
                    self.pos += 1;
                    if self.peek() == Some('=') {
                        self.pos += 1;
                        tokens.push(Token::EqualEqual);
                    } else {
                        tokens.push(Token::Equal);
                    }
                }
                Some('!') => {
                    self.pos += 1;
                    if self.peek() == Some('=') {
                        self.pos += 1;
                        tokens.push(Token::BangEqual);
                    } else {
                        tokens.push(Token::Illegal("unexpected !".into()));
                    }
                }
                Some('<') => {
                    self.pos += 1;
                    if self.peek() == Some('=') {
                        self.pos += 1;
                        tokens.push(Token::LessEqual);
                    } else {
                        tokens.push(Token::Less);
                    }
                }
                Some('>') => {
                    self.pos += 1;
                    if self.peek() == Some('=') {
                        self.pos += 1;
                        tokens.push(Token::GreaterEqual);
                    } else {
                        tokens.push(Token::Greater);
                    }
                }
                Some('-') => {
                    self.pos += 1;
                    if self.peek() == Some('>') {
                        self.pos += 1;
                        tokens.push(Token::Arrow);
                    } else {
                        tokens.push(Token::Minus);
                    }
                }
                Some('+') => { self.pos += 1; tokens.push(Token::Plus); }
                Some('*') => { self.pos += 1; tokens.push(Token::Star); }
                Some('/') => { self.pos += 1; tokens.push(Token::Slash); }
                Some('(') => { self.pos += 1; tokens.push(Token::LParen); }
                Some(')') => { self.pos += 1; tokens.push(Token::RParen); }
                Some('{') => { self.pos += 1; tokens.push(Token::LBrace); }
                Some('}') => { self.pos += 1; tokens.push(Token::RBrace); }
                Some(',') => { self.pos += 1; tokens.push(Token::Comma); }
                Some(';') => { self.pos += 1; tokens.push(Token::Semicolon); }
                Some(':') => { self.pos += 1; tokens.push(Token::Colon); }
                Some(c) => {
                    self.pos += 1;
                    tokens.push(Token::Illegal(format!("unexpected character: {}", c)));
                }
            }
        }

        tokens
    }
}
